### ISSUE TITLE
improve quotas not enabled behaviour

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#1869]
Under certain as yet undetermined circumstances quotas can become disabled. One consequence of this was a failure to mount the related shares upon the following boot / reboot (addressed in issue #1867 and related pr #1868). However there are additional consequences for running with a pool / subvol that has quotas not enabled. These include:

- [x] Inability to create new shares in that pool.
- [x] Inability to resize existing shares.
- [x] Apparent Inability to create a snapshot (error on trying but it appears)
- [x] Apparent Inability to delete a snapshot (error on trying but it get deleted)
~~[ ] Size reporting of the affected subvols is reported as 0 bytes.~~
- [x] Deal with quotas being re-enabled once shares etc have been created with default pqgroup -1/-1
- [x] Test pool import for ability to recreate/reset pqgroups
- [x] Test bootstrap command on quota enabled pool after having been used with quotas disabled ie -1/-1 entries in db.
- [x] Ensure all pqgroup's are unique while being robust to quote enable disable re-enabled cycle.
- [x] Add user feedback on Pool Quotas Enabled / Disabled 
- [x] Share created when quotas disabled get fresh pqgroup during re-size re-submission if done when quotas are enabled (no change needed).
- [x] Share with non existent pqgroup (due to quota disable removing all pqgroups from pool) will also be re-assigned a fresh pqgroup once quotas are re-enabled and a resize is performed.
- [x] remove all non existent pqgroups and re-assign fresh ones during boot if quotas enabled, already done for prior -1/-1 pqgroup shares.

**Comment by phillxnet** [comment]
After enabling debug via:
```

```
we have the following:
```
[08/Dec/2017 19:21:43] DEBUG [storageadmin.views.share_helpers:87] Share name = rock-ons-root.
[08/Dec/2017 19:21:43] DEBUG [storageadmin.views.share_helpers:89] Updating pre-existing same pool db share entry.
[08/Dec/2017 19:21:43] DEBUG [system.osi:104] Running command: /sbin/btrfs subvolume list /mnt2/rock-pool
[08/Dec/2017 19:21:43] DEBUG [system.osi:104] Running command: /sbin/btrfs qgroup show /mnt2/rock-pool/rock-ons-root
[08/Dec/2017 19:21:43] ERROR [system.osi:119] non-zero code(1) returned by command: ['/sbin/btrfs', 'qgroup', 'show', '/mnt2/rock-pool/rock-ons-root']. output: [''] error: ["ERROR: can't list qgroups: quotas not enabled", '']
[08/Dec/2017 19:21:43] DEBUG [storageadmin.views.share_helpers:87] Share name = test-share-2.
[08/Dec/2017 19:21:43] DEBUG [storageadmin.views.share_helpers:89] Updating pre-existing same pool db share entry.
[08/Dec/2017 19:21:43] DEBUG [system.osi:104] Running command: /sbin/btrfs subvolume list /mnt2/rock-pool
[08/Dec/2017 19:21:43] DEBUG [system.osi:104] Running command: /sbin/btrfs qgroup show /mnt2/rock-pool/test-share-2
[08/Dec/2017 19:21:43] ERROR [system.osi:119] non-zero code(1) returned by command: ['/sbin/btrfs', 'qgroup', 'show', '/mnt2/rock-pool/test-share-2']. output: [''] error: ["ERROR: can't list qgroups: quotas not enabled", '']
[08/Dec/2017 19:21:43] DEBUG [storageadmin.views.share_helpers:87] Share name = test-share.
[08/Dec/2017 19:21:43] DEBUG [storageadmin.views.share_helpers:89] Updating pre-existing same pool db share entry.
[08/Dec/2017 19:21:43] DEBUG [system.osi:104] Running command: /sbin/btrfs subvolume list /mnt2/rock-pool
[08/Dec/2017 19:21:43] DEBUG [system.osi:104] Running command: /sbin/btrfs qgroup show /mnt2/rock-pool/test-share
[08/Dec/2017 19:21:43] ERROR [system.osi:119] non-zero code(1) returned by command: ['/sbin/btrfs', 'qgroup', 'show', '/mnt2/rock-pool/test-share']. output: [''] error: ["ERROR: can't list qgroups: quotas not enabled", '']
...
```

**Comment by phillxnet** [comment]
post the following command:
```
btrfs quota enable /mnt2/rock-pool/
```
we have:
```
[08/Dec/2017 19:34:00] DEBUG [storageadmin.views.share_helpers:87] Share name = rock-ons-root.
[08/Dec/2017 19:34:00] DEBUG [storageadmin.views.share_helpers:89] Updating pre-existing same pool db share entry.
[08/Dec/2017 19:34:00] DEBUG [system.osi:104] Running command: /sbin/btrfs subvolume list /mnt2/rock-pool
[08/Dec/2017 19:34:00] DEBUG [system.osi:104] Running command: /sbin/btrfs qgroup show /mnt2/rock-pool/rock-ons-root
[08/Dec/2017 19:34:00] DEBUG [storageadmin.views.share_helpers:87] Share name = test-share-2.
[08/Dec/2017 19:34:00] DEBUG [storageadmin.views.share_helpers:89] Updating pre-existing same pool db share entry.
[08/Dec/2017 19:34:00] DEBUG [system.osi:104] Running command: /sbin/btrfs subvolume list /mnt2/rock-pool
[08/Dec/2017 19:34:00] DEBUG [system.osi:104] Running command: /sbin/btrfs qgroup show /mnt2/rock-pool/test-share-2
[08/Dec/2017 19:34:00] DEBUG [storageadmin.views.share_helpers:87] Share name = test-share.
[08/Dec/2017 19:34:00] DEBUG [storageadmin.views.share_helpers:89] Updating pre-existing same pool db share entry.
[08/Dec/2017 19:34:00] DEBUG [system.osi:104] Running command: /sbin/btrfs subvolume list /mnt2/rock-pool
[08/Dec/2017 19:34:00] DEBUG [system.osi:104] Running command: /sbin/btrfs qgroup show /mnt2/rock-pool/test-share
```

**Comment by phillxnet** [comment]
enable_quota(pool) called:
- when pool is created: add_pool() (fs/btrfs)
Not likely candidate as when pool was created in reproducer example install shares could be created and so quotas must have been enabled as otherwise we hit a current trigger when quota not enabled on pool.
- during a web-ui initiated import: _btrfs_disk_import() (storageadmin/views/disk.py)
Example reproducer had pool created from scratch so no import.
- during rescan but only for role.root (system) device: _update_disk_state() (storageadmin/views/disk.py)

The current suspicion is that we have a quota anomaly on certain pool instances where quota enabled status, such as is enacted upon pool creation, does not stick. Ie. after a reboot the pool has quotas disabled which is not currently supported in Rockstor: with the consequent limitations as noted in this issue. As such this issue may contain relevant info for issue #1592.

**Comment by phillxnet** [comment]
My current thinking on this issue is that we have a quota status not sticking issue on certain pool instances as my understanding is that Rockstor relies on the pool / subvols to maintain their quota status (enabled / disabled) over a reboot where as in the instances noted in this issue that is not happening: hence my recent links to plans to support disabling quotas or functioning with quotas disabled which we currently don't.

**Comment by phillxnet** [comment]
@schakrava I have found the following that may tie our disabled quotas to the recent docker update:

https://github.com/docker/for-linux/issues/78

Looks like a likely candidate to me.

**Comment by phillxnet** [comment]
I have transitioned this issue to improving our 'quota disabled' behaviour given we now have the culprit:
docker-ce disables quotas
So we might as well deal with it as best we can at least until this is fixed upstream.

I intend to open a more focused issue on "rock-ons-root pool quota disabled by docker-ce" in the near future.

**Comment by phillxnet** [comment]
User feedback elements re Pool Quotas status:

Column in Pools table:
![quota-status-on-pools-table](https://user-images.githubusercontent.com/2521585/33940311-600c0d0c-e006-11e7-8160-e6513553a524.png)

In Pool column - Shares table:
![quota-status-on-shares-table](https://user-images.githubusercontent.com/2521585/33940370-9a91725a-e006-11e7-9ec5-12388aacecd8.png)

On Pool details page:
![quota-status-on-pool-details-page](https://user-images.githubusercontent.com/2521585/33940387-a7eb45a2-e006-11e7-8331-d6b7948b7875.png)

On Shares details page re pool:
![quota-status-on-share-details-page](https://user-images.githubusercontent.com/2521585/33940407-c104566e-e006-11e7-89e4-948ef61ed135.png)

Text in Usage / Btrfs Usage columns in Shares table changed to show quotas dependency:
"Share content - uses Quotas"
"Share content inc snapshots - uses Quotas"

**Comment by phillxnet** [comment]
Rework of quota groups management in progress. Pull request to follow.

**Comment by phillxnet** [comment]
Please update the following forum thread with this issues resolution:
https://forum.rockstor.com/t/cant-create-new-share/4178

**Comment by schakrava** [comment]
Thanks for putting down all the details @phillxnet , very helpful!

We must(I really need to prioritize to do this myself) address the hard dependency on quotas in Rockstor. We need to investigate the current state of btrfs-quotas feature-set and perhaps let the user disable/enable quotas entirely or on per pool basis and also, as you indicated above, reflect the current quota enforcement to the user where it's appropriate.

**Comment by phillxnet** [comment]
@schakrava Yes, bar the size enhancement and usage reporting the code associated with this issue (pull request soon) seems to allow for all features I've tested so far to function with quotas disabled. Trying to work towards #1592 that we have had for a while.

I'll get this code in soon for review as it's so far successfully managed to re-make pqgroups with multiple successive quota disable / enable cycles. All pool quota state UI elements have been fairly aggressively tested it as I've been using it during the dev of the pqgroup management improvements.

**Comment by phillxnet** [comment]
Linking this now closed issue to:
"Feature: Quota rebuild script?" #1785
as our associated pr address (at least initially) the recreation of required rockstor native pqgroups (2015/*) as referenced in #1785 "... and re-create any qgroups it needs ..." and so represents a partial fix for #1785
